feat: Deprecate setup tools on init

bradbeam opened the pull request:
This introduces the ability to generate the installation manifests internally via redskyctl/kustomize
so we no longer need to rely on execing into the setup tools pod.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Deprecate setup tools on init

This introduces the ability to generate the installation manifests internally via redskyctl/kustomize
so we no longer need to rely on execing into the setup tools pod.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

jgustie commented on redskyctl/internal/commands/initialize/generator.go:
```diff
@@ -49,49 +52,22 @@ func NewGeneratorCommand(o *GeneratorOptions) *cobra.Command {
 func (o *GeneratorOptions) generate(ctx context.Context) error {
-	// Read the initial information from the configuration
-	r := o.Config.Reader()
-	ctrl, err := config.CurrentController(r)
```
The namespace for initialization needs to come out of the configuration, i.e. `ctrl.Namespace` is where `redsky-system` comes from.

jgustie commented on redskyctl/internal/commands/initialize/generator.go:
```diff
@@ -49,49 +52,22 @@ func NewGeneratorCommand(o *GeneratorOptions) *cobra.Command {
-	args = append(args, "--image", setup.Image)
-	// TODO We may need to overwrite this for offline clusters
-	args = append(args, "--image-pull-policy", setup.ImagePullPolicy)
+	fmt.Println(string(yamls))
```
This should `Fprintln` to `o.Out`.

jgustie commented on config/default/manager_patch.yaml:
```diff
@@ -0,0 +1,14 @@
+        envFrom:
+        - secretRef:
+            name: redsky-manager
+            optional: true
```
I would rather not have this moved into `config/default` as it that would include it into other downstream kustomizations.

bradbeam replied:
Should we try to inject this conditionally if there is a complete redskyctl config file?

jgustie replied:
Sure: this is here so we could have one static manifest for both the OSS and Enterprise editions, but having it optional has caused some unique timing related issues that are hard to debug.

bradbeam committed:
refactor: Generate a single assets file

This also removes the need for kustomize on host to generate the assets.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Swap common labels for transformers

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Dynamically inject envFrom for manager

This causes the manager deployment to reference the envFrom secret only
if the client credentials are present.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
test: Add tests for generate command

Add testing for generate install/trial commands.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

jgustie commented on Makefile:
```diff
@@ -38,8 +39,12 @@ manager: generate fmt vet
+	PULL_POLICY=${PULL_POLICY} \
+	REDSKYCTL_IMG=${REDSKYCTL_IMG} \
+	IMG=${IMG} \
+	goreleaser release --snapshot --skip-sign --rm-dist --debug
```
Is adding `--debug` here intentional?

jgustie replied:
I know we don't use this in CI, but in testing out the code signing, it definitely prints sensitive information.

bradbeam replied:
It shouldn't be there :)

jgustie commented on Makefile:
```diff
@@ -81,7 +87,7 @@ generate: controller-gen conversion-gen
 
 # Build the docker images
-docker-build-ci:
+docker-build-ci: manifests
```
Should we be committing the generated files like we do with the rest of the generated code so we do not need to regenerate manifests on every build?

bradbeam replied:
initially it was because I wanted the build version information from the CI run to get injected when we built the manifests. but after switching over to using kustomize, im not sure it's needed anymore. I'll double check and drop it if we don't need to

jgustie commented on redskyctl/internal/commands/initialize/generator.go:
```diff
@@ -49,49 +52,40 @@ func NewGeneratorCommand(o *GeneratorOptions) *cobra.Command {
-	// Overwrite the "redsky-system" namespace
-	args = append(args, "--env", "NAMESPACE="+ctrl.Namespace)
+	apiEnabled := false
+	if auth.Credential.TokenCredential != nil {
```
You should check `auth.Credential.ClientCredential` as well.

jgustie commented on cmd/generator/main.go:
```diff
@@ -0,0 +1,85 @@
+func main() {
+	var (
+		path          = flag.String("path", ".", "path of kustomization.yaml")
+		outputPackage = flag.String("package", "generated", "output package name")
```
This isn't configurable because of the reference to the `Asset` struct; you would need to conditionally add an `import` statement to keep it configurable.

bradbeam replied:
There's a dependency on having an Asset type defined, but the package name itself doesn't matter ( we're generating it for `package kustomize` ). We could probably move the `redskyctl/internal/kustomize/assets.go` bit into this command to generate the assets struct as well, but I'm not sure it's worth tossing all of that go code inside of a string literal or the like.
